File: FSharp_Library/ChineseService.fs

namespace Chinese

open System
open System.IO
open MySql.Data.MySqlClient
open System.Collections.Generic

//open MyTypes

//module ChineseService =

    //let filePath = @"C:\Users\chisi\Desktop\work\ChineseApp\FSharp_Library\SUBTLEX.utf8"
    //let wordsPath = @"C:\Users\chisi\Desktop\work\ChineseApp\FSharp_Library\allWords.utf8"

    //let allDetailedWords = new Dictionary<string, DetailedWord>()


    (*let getDetailedWordFromLine (line:string) =
        let tokens = line.Split('\t')
        {
            Simplified = tokens.[0];
            Length = tokens.[1];
            Pinyin = tokens.[2];
            PinyinInput = tokens.[3];
            WCount = tokens.[4];
            WMillion = tokens.[5];
            Log10W = tokens.[6];
            W_CD = tokens.[7];
            W_CD_percent = tokens.[8];
            Log10CD = tokens.[9];
            DominantPos = tokens.[10];
            DominantPosFreq = tokens.[11];
            AllPos = tokens.[12];
            AllPosFreq = tokens.[13];
            Definition = tokens.[14];
        }

    //delete first row file
    let buildAllDetailedWords() =
        for line in File.ReadAllLines(filePath) do
            let detailedWord = getDetailedWordFromLine(line)
            if not (allDetailedWords.ContainsKey(detailedWord.Simplified)) then
                allDetailedWords.Add(detailedWord.Simplified, detailedWord)*)

    
    //let allWords =
    //    let buildWordFromLine (line:string) =
    //        let token = line.Split('\t')
    //        {
    //            Traditional = token.[0];
    //            Simplified = token.[1];
    //            Pinyin = token.[2];
    //            Definitions = token.[3];
    //            Frequency = token.[4] |> int
    //        }
    //    File.ReadAllLines(wordsPath) |> Seq.map buildWordFromLine

    //let getAllDetailedWords() =
    //    buildAllDetailedWords()
    //    allDetailedWords

    //let getAllWords() =
    //    allWords

    //let getSortedByFrequency filteredWords =
    //    filteredWords |> Seq.sortBy (fun w -> w.Frequency) |> Seq.rev
    
    //let getEnglishResult (text:string) =
    //    allWords
    //   |> Seq.filter (fun w -> w.Definitions.Contains(text))
    //   |> getSortedByFrequency

    //let searchBySimplified (text:string) =
    //    allWords
    //    |> Seq.filter (fun w -> w.Simplified.Contains(text))
    //    |> getSortedByFrequency

    //TODO too C#-like
    (*let searchByPinyin (text:string) =
        let prons = text.Split(' ')

        let checkIfPinyinMatches (word:Word) =
            let wordProns = word.Pinyin.Split(' ')
            if prons.Length <> wordProns.Length then
                false
            else
                let mutable toInsert = true
                for i in 0 .. prons.Length-1 do
                    if  not (wordProns.[i].StartsWith (prons.[i])) then
                        toInsert <- false
                    if wordProns.[i].Length <> (prons.[i].Length + 1) then
                        toInsert <- false
                toInsert
        allWords
        |> Seq.filter checkIfPinyinMatches
        |> getSortedByFrequency*)

    //let getRandomWords() =
    //    let random = new Random()
    //    let result = new List<Word>()
        //for i in 0 .. 20 do
        //    let index = random.Next(allWordsAvecFrequency.Count)
        //    result.Add (allWordsAvecFrequency.[index])
    //    result

    //let getResultedWord (simpl:string) = 
    //    allWords
    //    |> Seq.filter (fun w -> w.Simplified = simpl)

    //todo clarify
    let getWordsFromSentence (sentence:string) =
        (*let mutable constructedWord = ""
        let mutable toInsert = new List<Word>()
        let mutable result = new List<Word>()
        for curr in sentence do
            let resultedWord = getResultedWord(constructedWord + curr.ToString()) |> Seq.toList //getResultedWord(constructedWord + curr)
            if resultedWord.Length > 0 then
                toInsert <- new List<Word>(resultedWord)
                constructedWord <- constructedWord + curr.ToString()
            else
                if toInsert.Count > 0 then
                    result.Add(toInsert.[0])
                toInsert <- new List<Word>(getResultedWord(curr.ToString())) //(getResultedWord(curr.ToString()))
                constructedWord <- curr.ToString()
        if toInsert.Count > 0 then
            result.Add(toInsert.[0]) //toInsert.ForEach(w => result.Add(w));
        result*)

        
        let mutable simpList = new List<string>()
        let mutable constructedWord = ""
        let mutable toInsert = ""
        let mutable resultList = new List<Word>()
        for curr in sentence do
            let wordToCheck = constructedWord + curr.ToString()
            if allDetailedWords.ContainsKey(wordToCheck) then
                toInsert <- allDetailedWords.[wordToCheck].Simplified
                constructedWord <- wordToCheck
            else
                if toInsert <> "" then
                    simpList.Add(toInsert)
                    toInsert <- ""
                toInsert <- curr.ToString()
        if toInsert <> "" then
            simpList.Add(toInsert)

        for simp in simpList do
            let words = allWords |> Seq.filter (fun w -> w.Simplified = simp)
            for word in words do
                resultList.Add(word)
        resultList
        
       




        

File: WPF_Program/Logic/ChineseService.cs
using ChineseAppWPF.Models;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.RegularExpressions;

namespace ChineseAppWPF.Logic
{
    public static partial class ChineseService
    {
        private static readonly List<Word> allWords = BuildAllWords();
        private static readonly Dictionary<string, DetailedWord> allDetailedWords = BuildAllDetailedWords();
        private static readonly HashSet<string> wordsSet = BuildAllWordsSet();
        private static readonly HashSet<char> characterSet = BuildCharacterSet();
        private static readonly HashSet<string> onlyTraditionalSet = BuildOnlyTraditionalSet();
        private static readonly Dictionary<char, int> strokesDict = BuildStrokesDict();

        public static List<Word> GetAllWords() => allWords;

        public static Dictionary<string, DetailedWord> GetAllDetailedWords() => allDetailedWords;


        public static bool ContainsInsensitive(this string source, string value)
        {
            return source?.IndexOf(value, StringComparison.InvariantCultureIgnoreCase) >= 0;
        }

        public static IEnumerable<Word> GetEnglishResult(string text)
        {
            return allWords.AsParallel()
                           .Where(w => w.Definitions.ContainsInsensitive(text));
        }

        public static bool ContainsRegexPattern(string text)
        {
            return text.Contains('^') ||
                   text.Contains('*') ||
                   text.Contains('.') ||
                   text.Contains('$') ||
                   text.Contains('+') ||
                   text.Contains('?');
        }

        public static IEnumerable<Word> SearchBySimplified(string text)
        {
            if (ContainsRegexPattern(text))
            {
                Regex rgx = new Regex(@text);
                ChineseSystem writingSystem = GetWritingSystem(text);

                if (writingSystem == ChineseSystem.Simplified)
                {
                    return allWords.AsParallel()
                                   .Where(w => rgx.IsMatch(w.Simplified));
                }
                else
                {
                    return allWords.AsParallel()
                                   .Where(w => rgx.IsMatch(w.Traditional));
                }
            }
            else
            {
                ChineseSystem writingSystem = GetWritingSystem(text);

                if (writingSystem == ChineseSystem.Simplified)
                {
                    return allWords.AsParallel()
                                   .Where(w => w.Simplified.Contains(text));
                }
                else
                {
                    return allWords.AsParallel()
                                   .Where(w => w.Traditional.Contains(text));
                }
            }
        }

        public static IEnumerable<Word> SearchByPinyin(string text)
        {
            if (string.IsNullOrEmpty(text))
                return new List<Word>();
            string[] prons = text.Split(' ');
            bool CheckIfPinyinMatches(Word word)
            {
                string[] wordProns = word.Pinyin.Split(' ');
                if (prons.Length != wordProns.Length)
                    return false;
                for (int i = 0; i <= prons.Length - 1; i++)
                {
                    if (!wordProns[i].StartsWith(prons[i]))
                        return false;
                    if (wordProns[i].Length != (prons[i].Length + 1))
                        return false;
                }
                return true;
            }
            return allWords.AsParallel()
                           .Where(CheckIfPinyinMatches);
        }

        public static IEnumerable<Word> GetRandomWords()
        {
            Random random = new Random();
            List<Word> result = new List<Word>();
            for (int i = 0; i < 20; i++)
            {
                int index = random.Next(allWords.Count);
                result.Add(allWords[index]);
            }
            return result;
        }

        internal static bool IsCharacter(char character)
        {
            return characterSet.Contains(character);
        }

        internal static bool IsPunctuation(string word)
        {
            return word == "," || word == "，" ||
                   word == "." || word == "。" ||
                   word == "?" || word == "？" ||
                   word == "!" || word == "！";
        }

        internal static bool WordExists(string simplOfWord)
        {
            if (IsPunctuation(simplOfWord))
                return true;
            return wordsSet.Contains(simplOfWord);
        }

        public static IEnumerable<Word> GetAllWordsFrom(char character)
        {
            string ch = character.ToString();
            if (GetWritingSystem(ch) == ChineseSystem.Simplified)
            {
                return allWords.Where(w => w.Simplified == ch);
            }
            else
            {
                return allWords.Where(w => w.Traditional == ch);
            }
        }

        public static IEnumerable<Word> GetAllWordsFrom(IEnumerable<string> simpList)
        {
            ChineseSystem writingSystem = GetWritingSystem(simpList);

            if (writingSystem == ChineseSystem.Simplified)
            {
                return from simp in simpList
                       from w in allWords
                       where w.Simplified == simp
                       select w;
            }
            else
            {
                return from simp in simpList
                       from w in allWords
                       where w.Traditional == simp
                       select w;
            }
        }

        public static ChineseSystem GetWritingSystem(string text)
        {
            if (onlyTraditionalSet.Contains(text))
                return ChineseSystem.Traditional;
            return ChineseSystem.Simplified;
        }

        public static ChineseSystem GetWritingSystem(IEnumerable<string> simpList)
        {
            foreach (string s in simpList)
            {
                if (GetWritingSystem(s) == ChineseSystem.Traditional)
                {
                    return ChineseSystem.Traditional;
                }
            }
            return ChineseSystem.Simplified;
        }

        public static IEnumerable<string> GetSimplifiedWordsFromSentence(string sentence)
        {
            int i = 0;
            int j = sentence.Length;
            List<string> result = new List<string>();

            while (i < j)
            {
                string possible = sentence[i..j];
                if (WordExists(possible))
                {
                    result.Add(possible);
                    i = j;
                    j = sentence.Length;
                }
                else
                {
                    j--;
                }
            }
            return result;
        }

        public static List<Word> GetWordsWithSimplified(string simplified)
        {
            return allWords.Where(w => w.Simplified == simplified).ToList();
        }

        public static List<Word> SearchWordsInside(string simplified)
        {
            List<Word> result = new List<Word>();

            for (int i = 0; i < simplified.Length; i++)
            {
                for (int j = i + 1; j <= simplified.Length; j++)
                {
                    string possible = simplified[i..j];
                    if (WordExists(possible))
                    {
                        result.AddRange(GetWordsWithSimplified(possible));
                    }
                }
            }
            result.Reverse();
            return result.OrderBy(w => w.Simplified.Length).Reverse().ToList();
        }
    }
}